practice-tool/src/widgets/character_stats.rs

```diff
@@ -1,19 +1,34 @@
+use std::borrow::BorrowMut;
 use libeldenring::prelude::*;
 use practice_tool_core::key::Key;
 use practice_tool_core::widgets::stats_editor::{Datum, Stats, StatsEditor};
 use practice_tool_core::widgets::Widget;
 
 #[derive(Debug)]
 struct CharacterStatsEdit {
-    ptr: PointerChain<CharacterStats>,
+    stats_ptr: PointerChain<CharacterStats>,
+    blessings_ptr: Option<PointerChain<CharacterBlessings>>,
     stats: Option<CharacterStats>,
+    blessings: Option<CharacterBlessings>,
 }
 
 impl Stats for CharacterStatsEdit {
-    fn data(&mut self) -> Option<impl Iterator<Item = Datum>> {
-        self.stats.as_mut().map(|s| {
-            [
-                Datum::int("Level", &mut s.level, 1, i32::MAX),
+    fn data(&mut self) -> Option<impl Iterator<Item=Datum>> {
+        if let Some(s) = self.stats.as_mut() {
+            let has_blessings: bool;
+            let b: &mut CharacterBlessings;
+            if let Some(blessings) = self.blessings.as_mut() {
+                has_blessings = true;
+                b = blessings;
+            } else {
+                has_blessings = false;
+                unsafe {
+                    static mut BLESSINGS_NONE: CharacterBlessings = CharacterBlessings { scadutree: 0, revered_spirit_ash: 0 };
+                    b = BLESSINGS_NONE.borrow_mut();
+                }
+            }
```

Comment:
The semantics here are a bit confusing, significantly complicate the method, and use `unsafe` as a wiorkaround which we should avoid at all costs.

The following pseudocode [chaining](https://doc.rust-lang.org/std/iter/struct.Chain.html) iterators would be much cleaner:

```rs
self.stats.as_mut().map(|s| {
    let base_data = [ /* ...the result of the previous closure */ ].into_iter();

    if let Some(b) = self.blessings.as_mut() {
        base_data.chain([
            Datum::byte("Scadutree Blessing", &mut b.scadutree, 0, 20),
            Datum::byte("Revered Spirit Ash", &mut b.revered_spirit_ash, 0, 20),
        ].into_iter())
    } else {
        base_data
    }
```

Reply:
This would imply compiler error due to different lengths in returned Iterrator from 2 branches

Reply:
Oh, right, that makes sense yeah.